dataset/shape.cpp

```diff
@@ -141,4 +141,232 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
+///    variable's dims, broadcast
+/// 3. If none of the variables's dimensions are contained, no broadcast
+Variable maybe_broadcast(const VariableConstView &var,
+                         const Dimensions &from_dims) {
+  // TODO: do we need a std::move?
+
+  const auto &var_dims = var.dims();
+
+  bool vardim_found_in_fromdims = false;
+  for (const auto dim : var_dims.labels())
+    if (from_dims.contains(dim))
+      vardim_found_in_fromdims = true;
+
+  bool vardims_contains_all_fromdims = true;
+  for (const auto dim : from_dims.labels())
+    if (!var_dims.contains(dim))
+      vardims_contains_all_fromdims = false;
+
+  // if (need_broadcast && !var_dims.contains(from_dims)) {
+  if (vardim_found_in_fromdims && !vardims_contains_all_fromdims) {
+    Dimensions broadcast_dims;
+    for (const auto dim : var_dims.labels())
+      if (!from_dims.contains(dim))
+        broadcast_dims.addInner(dim, var_dims[dim]);
+      else
+        for (const auto lab : from_dims.labels())
+          // Need to check if the variable contains that dim, and use the
+          // variable shape in case we have a bin edge.
+          if (var_dims.contains(lab))
+            broadcast_dims.addInner(lab, var_dims[lab]);
+          else
+            broadcast_dims.addInner(lab, from_dims[lab]);
+    return broadcast(var, broadcast_dims);
+  } else {
+    return Variable(std::move(var));
+  }
+}
+
+// In the case of bin edges, we perform a series of slices and concatenations
+// to carry out the reshape. This is not the most efficient for large data,
+// but it makes the code simple.
+Variable slice_and_stack(const VariableConstView &var, const Dim from_dim,
+                         const Dimensions &to_dims) {
+  const auto labels = to_dims.labels();
+  scipp::index step;
+  Variable reshaped;
+  auto copy = Variable(var);
+  for (uint64_t l = 0; l < labels.size() - 1; ++l) {
+    step = copy.dims()[from_dim] / to_dims.shape()[l];
+    reshaped = Variable(copy.slice({from_dim, 0, step + 1}));
+    for (int32_t i = 1; i < to_dims[labels[l]]; ++i)
+      reshaped = concatenate(
+          reshaped, copy.slice({from_dim, i * step, (i + 1) * step + 1}),
+          labels[l]);
+    // Update copy
+    copy = reshaped;
+  }
+  reshaped.rename(from_dim, labels.back());
+  return reshaped;
+}
+
+Variable slice_and_unstack(const VariableConstView &var,
+                           const Dimensions &from_dims, const Dim to_dim) {
+  const auto labels = from_dims.labels();
+  Variable reshaped;
+  auto copy = Variable(var);
+  for (uint64_t l = 0; l < labels.size() - 1; ++l) {
+    if (copy.dims().contains(labels[l])) {
+      reshaped = Variable(copy.slice({labels[l], 0}));
+      for (int32_t i = 1; i < from_dims.shape()[l]; ++i)
+        reshaped = join_edges(VariableView(reshaped),
+                              copy.slice({labels[l], i}), labels.back());
+      // Update copy
+      copy = reshaped;
+    }
```

Comment:
Maybe a better and faster way to do this would be to slice out the last slice along all the `from_dims`, reshape/flatten that and then add/concat the very last point?

Reply:
I went for another approach in the end, that was similar to the one you suggested for flattening the coord